File: code/ex60.py
import math

class Solution:
    def getPermutation(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """

        num = [(i + 1) for i in range(n)]

        r = ''

        k -= 1

        while n > 0:
            n -= 1

            index = k // (math.factorial(n))
            r += str(num[index])
            k = k % (math.factorial(n))

            #Remove index item
            num.remove(num[index])

        return r

sol = Solution()
print(sol.getPermutation(3, 3))

File: code/ex60.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        k--;
        string res = "";
        int total = factorial(n);    // Number of digit candidates, total choices on current digit
        unordered_set<int> available;    // All available digit candidates
        for (int i = 1; i <= n; i++)    available.insert(i);    // Get all candidates
        
        for (int d = 0; d < n; d++) {
            int loop = total / available.size();
            int i = k / loop;
            k = k % loop;    // The kth item in current loop
            // Get the actural digit
            int cnt = 0;
            for (int digit = 1; digit <= n; digit++) {
                if (available.count(digit) == 0)    continue;
                if (cnt == i)    {
                    res += to_string(digit);
                    available.erase(digit);
                    break;
                }
                cnt++;
            }
            total = loop; 
        }
        
        return res;
    }
    
private:
    int factorial(int n) {
        int res = 1;
        for (int i = n; i > 1; i--)
            res *= i;
        return res;
    }
};